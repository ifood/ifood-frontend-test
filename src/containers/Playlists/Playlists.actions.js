const REQUEST = 'REQUEST';
const SUCCESS = 'SUCCESS';
const FAILURE = 'FAILURE';

export const PLAYLISTS = createRequestTypes('PLAYLISTS');
export const FILTERS_METADATA = createRequestTypes('FILTERS_METADATA');

export function action(type, payload = {}) {
  return { type, ...payload };
}

export function createRequestTypes(base) {
  return [REQUEST, SUCCESS, FAILURE].reduce((acc, type) => {
    acc[type] = `${base}_${type}`;
    return acc;
  }, {});
}

export const playlists = {
  request: filters => action(PLAYLISTS[REQUEST], { filters }),
  success: response => action(PLAYLISTS[SUCCESS], { data: response }),
  failure: error => action(PLAYLISTS[FAILURE], { error }),
};

export const filtersMetadata = {
  request: () => action(FILTERS_METADATA[REQUEST]),
  success: metadata => action(FILTERS_METADATA[SUCCESS], { metadata }),
  failure: error => action(FILTERS_METADATA[FAILURE], { error }),
};
