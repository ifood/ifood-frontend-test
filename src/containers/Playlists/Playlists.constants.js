export const PLAYLISTS_URL = 'https://api.spotify.com/v1/browse/featured-playlists';
export const FILTER_METADATA_URL = 'http://www.mocky.io/v2/5a25fade2e0000213aa90776';
export const REFRESH_INTERVAL = 30000;

export default {
  PLAYLISTS_URL,
  FILTER_METADATA_URL,
  REFRESH_INTERVAL,
};
