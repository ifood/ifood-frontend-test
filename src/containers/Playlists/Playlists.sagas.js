import { call, put, takeLatest } from 'redux-saga/effects';
import Api from './Playlists.api';
import {
  PLAYLISTS,
  FILTERS_METADATA,
  playlists,
  filtersMetadata,
} from './Playlists.actions';

export function* fetchPlaylists(action) {
  try {
    const response = yield call(Api.fetchPlaylists, action.filters);
    yield put(playlists.success(response));
  } catch (e) {
    yield put(playlists.failure(e.message));
  }
}

export function* fetchFiltersMetadata() {
  try {
    const metadata = yield call(Api.fetchFiltersMetadata);
    yield put(filtersMetadata.success(metadata));
  } catch (e) {
    yield put(filtersMetadata.failure(e.message));
  }
}

function* playListsSaga() {
  yield takeLatest(PLAYLISTS.REQUEST, fetchPlaylists);
  yield takeLatest(FILTERS_METADATA.REQUEST, fetchFiltersMetadata);
}

export default playListsSaga;
