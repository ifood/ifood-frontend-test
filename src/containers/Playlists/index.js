import React, { useState, useEffect } from "react";
import { getPlaylists } from "services/api";
import { ReactComponent as YourSvg } from "assets/logo.svg";
import PlaylistCard from "components/PlaylistCard";

export default function Playlists() {
  const [playlists, setPlaylists] = useState([]);
  const [isLoading, setLoading] = useState(false);

  const loadPlaylists = async () => {
    try {
      setLoading(true);
      const { data } = await getPlaylists();
      if (data.playlists && Array.isArray(data.playlists.items)) {
        setPlaylists(data.playlists.items);
      }
    } catch (e) {
      console.log(e);
      // TODO: Show error message on screen
    }
    setLoading(false);
  };

  useEffect(() => {
    loadPlaylists();
  }, []);

  return (
    <div className="playlists">
      <YourSvg />
      <div className="playlists__items">
        {playlists.map((playlist) => {
          return <PlaylistCard playlist={playlist} />;
        })}
      </div>
    </div>
  );
}
