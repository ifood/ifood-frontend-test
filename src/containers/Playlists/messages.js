import { defineMessages } from "react-intl";

export const scope = "containers.Playlists";

export default defineMessages({
  hideFilters: {
    id: `${scope}.hideFilters`,
    defaultMessage: "Hide filters",
  },
  expandFilters: {
    id: `${scope}.expandFilters`,
    defaultMessage: "Expand filters",
  },
  contentAriaLAbel: {
    id: `${scope}.contentAriaLAbel`,
    defaultMessage: "Playlist explorer",
  },
});
