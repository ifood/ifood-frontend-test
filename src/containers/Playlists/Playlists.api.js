import axios from 'axios';
import { PLAYLISTS_URL, FILTER_METADATA_URL } from './Playlists.constants';

class PlaylistsApi {
  static fetchPlaylists = async (params) => {
    const accessToken = localStorage.getItem('spotifood-access-token');
    const options = {
      headers: { Authorization: `Bearer ${accessToken}` },
      params,
    };

    const response = await axios.get(PLAYLISTS_URL, options);
    return response.data.playlists.items;
  }

  static fetchFiltersMetadata = async () => {
    const response = await axios.get(FILTER_METADATA_URL);
    return response.data.filters;
  }
}

export default PlaylistsApi;
