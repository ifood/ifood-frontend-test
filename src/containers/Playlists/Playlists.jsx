import React, { Component, Fragment } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import pickBy from 'lodash/pickBy';
import CircularProgress from '@material-ui/core/CircularProgress';
import Header from '../../components/Header';
import Filters from './components/Filters';
import PlaylistGrid from './components/PlaylistGrid';
import {
  playlists as playlistActions,
  filtersMetadata as filtersMetadataActions,
} from './Playlists.actions';
import { REFRESH_INTERVAL } from './Playlists.constants';

const Container = styled.div`
  flex-grow: 1;
  padding: 25px;
`;

const ErrorMessage = styled.div`
  margin-top: 50px;
  width: 100%;
  text-align: center;
  color: #dc3545;
  font-size: 16px;
`;

const SpinnerContainer = styled.div`
  margin-top: 50px;
  width: 100%;
  text-align: center;
  color: #bc2026;
`;

class Playlists extends Component {
  state = {
    filteredPlaylists: [],
    searchTerm: '',
  }

  componentDidMount() {
    const { fetchPlayLists, fetchFiltersMetadata } = this.props;
    const intervalId = setInterval(fetchPlayLists, REFRESH_INTERVAL);
    fetchPlayLists();
    fetchFiltersMetadata();
    this.setState({ intervalId });
  }

  componentWillReceiveProps(nextProps) {
    const { searchTerm } = this.state;
    const { playlists } = this.props;
    if (nextProps.playlists !== playlists) {
      this.setState({
        filteredPlaylists: this.filterByText(nextProps.playlists, searchTerm),
      });
    }
  }

  componentWillUnmount() {
    const { intervalId } = this.state;
    clearInterval(intervalId);
  }

  filterByText = (playlists, searchTerm) => {
    if (searchTerm === '') {
      return playlists;
    }

    return playlists.filter((playlist) => {
      return playlist.name.toLowerCase().includes(searchTerm.toLowerCase());
    });
  }

  onFiltersChange = (filters) => {
    const { fetchPlayLists } = this.props;
    const validFilters = pickBy(filters, value => value !== '');
    fetchPlayLists(validFilters);
  }

  onSearchChange = (searchTerm) => {
    const { playlists } = this.props;
    const filteredPlaylists = playlists.length
      ? this.filterByText(playlists, searchTerm)
      : playlists;

    this.setState({
      searchTerm,
      filteredPlaylists,
    });
  }

  renderPlaylistGrid = () => {
    const { filteredPlaylists } = this.state;
    const { isLoadingPlaylists } = this.props;

    if (isLoadingPlaylists) {
      return (
        <SpinnerContainer>
          <CircularProgress />
        </SpinnerContainer>
      );
    }

    if (!filteredPlaylists.length) {
      return (
        <ErrorMessage>
          There are no playlists for these filters, please change them and try again
        </ErrorMessage>
      );
    }
    return <PlaylistGrid playlists={filteredPlaylists} />;
  }

  render() {
    const { metaFilters } = this.props;

    return (
      <Fragment>
        <Header title="Spotifood" />
        <Container>
          <Filters
            metaFilters={metaFilters}
            onChange={this.onFiltersChange}
            onSearch={this.onSearchChange}
          />
          {this.renderPlaylistGrid()}
        </Container>
      </Fragment>
    );
  }
}

Playlists.propTypes = {
  fetchPlayLists: PropTypes.func.isRequired,
  fetchFiltersMetadata: PropTypes.func.isRequired,
  metaFilters: PropTypes.array,
  playlists: PropTypes.array,
  isLoadingPlaylists: PropTypes.bool,
};

Playlists.defaultProps = {
  metaFilters: null,
  playlists: undefined,
  isLoadingPlaylists: false,
};

const mapStateToProps = (state) => {
  return {
    playlists: state.playlistsContainer.playlists.data,
    isLoadingPlaylists: state.playlistsContainer.playlists.loading,
    metaFilters: state.playlistsContainer.filters.metadata,
    isLoadingFilters: state.playlistsContainer.filters.loading,
  };
};

export { Playlists };

export default connect(mapStateToProps, {
  fetchPlayLists: playlistActions.request,
  fetchFiltersMetadata: filtersMetadataActions.request,
})(Playlists);
