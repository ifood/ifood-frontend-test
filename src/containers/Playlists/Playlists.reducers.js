import { combineReducers } from 'redux';
import { PLAYLISTS, FILTERS_METADATA } from './Playlists.actions';

export const playlistsData = (state = {}, action) => {
  switch (action.type) {
    case PLAYLISTS.REQUEST:
      return Object.assign({}, { ...state, loading: true });
    case PLAYLISTS.SUCCESS:
      return Object.assign({}, { ...state, loading: false, data: action.data });
    case PLAYLISTS.FAILURE:
      return Object.assign({}, {
        ...state,
        loading: false,
        error: action.error,
        data: [],
      });
    default:
      break;
  }

  return state;
};

export const filtersMetadata = (state = {}, action) => {
  switch (action.type) {
    case FILTERS_METADATA.REQUEST:
      return Object.assign({}, { ...state, loading: true });
    case FILTERS_METADATA.SUCCESS:
      return Object.assign({}, { ...state, loading: false, metadata: action.metadata });
    case FILTERS_METADATA.FAILURE:
      return Object.assign({}, { ...state, loading: false, error: action.error });
    default:
      break;
  }

  return state;
};

export default combineReducers({
  playlists: playlistsData,
  filters: filtersMetadata,
});
