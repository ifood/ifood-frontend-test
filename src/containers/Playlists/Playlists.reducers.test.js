import { playlistsData, filtersMetadata } from './Playlists.reducers';
import { PLAYLISTS, FILTERS_METADATA } from './Playlists.actions';

describe('Playlists.reducers', () => {
  const mocks = {};

  beforeEach(() => {
    mocks.action = {};
  });

  describe('playlistsData()', () => {
    it('should return the correct new state for PLAYLISTS.REQUEST action', () => {
      mocks.action = {
        type: PLAYLISTS.REQUEST,
      };
      const newState = playlistsData({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });

    it('should return the correct new state for PLAYLISTS.SUCCESS action', () => {
      mocks.action = {
        type: PLAYLISTS.SUCCESS,
        data: ['1', '2'],
      };
      const newState = playlistsData({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });

    it('should return the correct new state for PLAYLISTS.FAILURE action', () => {
      mocks.action = {
        type: PLAYLISTS.FAILURE,
        error: 'Some error',
      };
      const newState = playlistsData({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });
  });

  describe('filtersMetadata()', () => {
    it('should return the correct new state for FILTERS_METADATA.REQUEST action', () => {
      mocks.action = {
        type: FILTERS_METADATA.REQUEST,
      };
      const newState = filtersMetadata({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });

    it('should return the correct new state for FILTERS_METADATA.SUCCESS action', () => {
      mocks.action = {
        type: FILTERS_METADATA.SUCCESS,
        metadata: ['1', '2'],
      };
      const newState = filtersMetadata({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });

    it('should return the correct new state for FILTERS_METADATA.FAILURE action', () => {
      mocks.action = {
        type: FILTERS_METADATA.FAILURE,
        error: 'Some error',
      };
      const newState = filtersMetadata({ something: 'already there' }, mocks.action);
      expect(newState).toMatchSnapshot();
    });
  });
});
