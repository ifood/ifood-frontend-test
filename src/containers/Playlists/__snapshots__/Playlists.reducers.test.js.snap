// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlists.reducers filtersMetadata() should return the correct new state for FILTERS_METADATA.FAILURE action 1`] = `
Object {
  "error": "Some error",
  "loading": false,
  "something": "already there",
}
`;

exports[`Playlists.reducers filtersMetadata() should return the correct new state for FILTERS_METADATA.REQUEST action 1`] = `
Object {
  "loading": true,
  "something": "already there",
}
`;

exports[`Playlists.reducers filtersMetadata() should return the correct new state for FILTERS_METADATA.SUCCESS action 1`] = `
Object {
  "loading": false,
  "metadata": Array [
    "1",
    "2",
  ],
  "something": "already there",
}
`;

exports[`Playlists.reducers playlistsData() should return the correct new state for PLAYLISTS.FAILURE action 1`] = `
Object {
  "data": Array [],
  "error": "Some error",
  "loading": false,
  "something": "already there",
}
`;

exports[`Playlists.reducers playlistsData() should return the correct new state for PLAYLISTS.REQUEST action 1`] = `
Object {
  "loading": true,
  "something": "already there",
}
`;

exports[`Playlists.reducers playlistsData() should return the correct new state for PLAYLISTS.SUCCESS action 1`] = `
Object {
  "data": Array [
    "1",
    "2",
  ],
  "loading": false,
  "something": "already there",
}
`;
